### PR TITLE
[web] Improve core/Page component and change how Agama layout is handled

### PR DESCRIPTION
**Pull request by dgdavid**
## The main change

As the project evolves, some decisions have been taken in regard to the elements' placement in the UI. Although nothing is set in stone, the truth is that these changes have been steady for several months already. However, the UI internals haven't been updated accordingly to them, keeping a complexity that comes from the early days of Agama when everything was less defined.

It is the case of the core/Layout component and its slots, managed by the react-teleporter dependency. They were there because of the need to _influence_ the layout from its children, which fortunately is no longer the case. E.g., it was needed to "teleport" a navigation link from the page or page's section to the sidebar, something that was called "contextual actions". Later, we realized that it was better to have a kind of page menu accessible from the header. What is more, the Sidebar ended up being relegated for holding only Agama stuff, which makes it kind of _static_.

Thus, core/Layout and its _teleportable_ slots does not make sense anymore. At this time they just provide a worst DX and a no longer justifiable use of React Portals. Moreover, since core/Page has become a cornerstone while it comes to Agama UI, it looks reasonable to build the layout from there. Doesn't matter if in the future it's needed to manage different page layouts, they can be extracted and _injected_ via core/Page props or using another technique. About performance, it turns out that _rendering everything_ when user navigates from one page to another does not have a so negative impact. Remember that React (the UI library Agama is using) [only changes the DOM nodes if there’s a difference between renders](https://react.dev/learn/render-and-commit#step-3-react-commits-changes-to-the-dom). Furthermore, asking for _re-renders_ on user navigation is not as different from a basic client/server web application: when the user navigates, the server sends back new content and usually there is no performance penalty since browsers know how to do their work efficiently even when for this example they have to fully _updates_ the [DOM](https://dom.spec.whatwg.org/). In any case, _by good luck_ we're constantly trying to keep the DOM size as minimum as possible, see https://web.dev/articles/dom-size-and-interactivity.

## Other things

Additionally, some other changes have been done as part of this set of changes, some of them needed to accomplish the main goal described above. Take a look at the commits to know more.

## Notes

From https://github.com/openSUSE/agama/pull/925#discussion_r1427405284:

> **A note about using `data-` attributes for styling**
> 
> I started using them for _exceptions_, as [suggested by CUBE](https://cube.fyi/exception.html#exception) methodology which I'd like to follow (WIP). But lately I found myself using them more to make styles easier to understand and, at some level, improve DX.
> 
> I'm aware that CSS _is_ mostly about classes selectors and they are likely to be more performant than data- attributes when used for styling. Though I haven't tested it myself, the potential performance penalty appears to be insignificant. See https://webdev.tips/styling-using-data-attributes#heading-performance-considerations, https://ecss.benfrain.com/appendix1.html, and https://gomakethings.com/how-performant-are-data-attributes-as-selectors/
> 
> Keeping a balance is the key here. Making use of them when it seems reasonable, and most importantly, trying **not to abuse them**. It's actually **my** goal to **use as few selectors as possible**, regardless of their type.
> 
> Of course, I can be wrong when taking a decision of adding or removing a selector. And that's why I'm adding this note, to let you know what's the main intention behind my _CSS movements_ and to encourage you to join me in the joy of exploring how to make the Agama CSS great and **minimal** at the same time :)
> 
> Also, I've to say that I'm not afraid to make mistakes here. They give me/us the opportunity to learn and improve. Furthermore, CSS **is easy to change and fix**.
> 
> BTW, when looking for evidences about data- selectors performance I found an article talking about [Class-less CSS architecture](https://jacobxperez.github.io/blog/post/css/class-less-architecture/)  :innocent:  

---

Related to https://trello.com/c/h367wPYP (internal link)

### CODE REVIEW

File: web/src/components/core/Sidebar.jsx [hunk 1]
```diff
@@ -130,8 +130,9 @@ export default function Sidebar ({ children, isOpen, onClose = noop }) {
     <aside
       id="global-options"
       ref={asideRef}
-      className="wrapper sidebar"
       aria-label={_("Installer Options")}
+      data-type="agama/sidebar"
+      data-layout="agama/base"
```
Comment (dgdavid) [inline web/src/components/core/Sidebar.jsx:135]: **A note about using `data-` attributes for styling**

---

I started using them for _exceptions_, as [suggested by CUBE](https://cube.fyi/exception.html#exception) methodology which I'd like to follow (WIP). But lately I found myself using them more to make styles easier to understand and, at some level, improve DX.

I'm aware that CSS _is_ mostly about classes selectors and they are likely to be more performant than data- attributes when used for styling. Though I haven't tested it myself, the potential performance penalty appears to be insignificant. See https://webdev.tips/styling-using-data-attributes#heading-performance-considerations, https://ecss.benfrain.com/appendix1.html, and https://gomakethings.com/how-performant-are-data-attributes-as-selectors/

Keeping a balance is the key here. Making use of them when it seems reasonable, and most importantly, trying not to abuse them. It's actually **my** goal to **use as few selectors as possible**, regardless of their type.

Of course, I can be wrong when taking a decision of adding or removing a selector. And that's why I'm adding this note, to let you know what's the main intention behind my _CSS movements_ and to encourage you to join me in the joy of exploring how to make the Agama CSS great and **minimal** at the same time :)

Also, I've to say that I'm not afraid to make mistakes here. They give me/us the opportunity to learn and improve. Furthermore, CSS **is easy to change and fix**.

BTW, when looking for evidences about data- selectors performance I found an article talking about [Class-less CSS architecture](https://jacobxperez.github.io/blog/post/css/class-less-architecture/)  :innocent:

File: web/src/components/core/DevelopmentInfo.jsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+import React from "react";
+import { _ } from "~/i18n";
+
+/**
+ * Displays additional info when running in a development server
+ * @component
+ */
+export default function DevelopmentInfo () {
```
Comment (dgdavid) [inline web/src/components/core/DevelopmentInfo.jsx:29]: Probably this deserves a better name (apart from a more complete test). And a better place, maybe it's time to create a _components/development/_ namespace to leave there components to be used only in development mode.

I'm thinking in an area in the sidebar for putting things related to Agama but only relevant when running in development mode.

File: web/src/components/core/If.jsx [hunk 3]
```diff
@@ -20,12 +20,43 @@
...
+ *     <Card>
+ *       <Title>Loading data</Title>
+ *       <Body>
+ *         <If condition={loadingStorage} then={<LoadingStorageData />} else={<StorageLoaded} />
+ *         <If condition={loadingNetwork} then={<LoadingNetwork />} else={<NetworkLoaded} />
+ *       </Body>
+ *     </Card>
+ *   );
+ *   ...
+ *
```
Comment (dgdavid) [inline web/src/components/core/If.jsx:59]: Better documentation and examples are more than welcome :innocent:

File: web/src/components/core/If.jsx [hunk 4]
```diff
@@ -20,12 +20,44 @@
...
+ *       <Title>Loading data</Title>
+ *       <Body>
+ *         <If condition={loading} then={<Spinner />} />
+ *         <If condition={loadingStorage} then={<LoadingStorageData />} else={<StorageLoaded />} />
+ *         <If condition={loadingNetwork} then={<LoadingNetwork />} else={<NetworkLoaded />} />
+ *       </Body>
+ *     </Card>
+ *   );
+ *   ...
+ *
```
Comment (dgdavid) [inline web/src/components/core/If.jsx:60]: Better documentation and examples are more than welcome :innocent:

**Review by imobachgs — COMMENTED** [review]
I like the direction you set in this pull request. Everything looks much simpler now. Great job!

By the way, thanks a ton for so informative commit descriptions. Something to learn from you.

File: web/src/components/core/Page.jsx [hunk 5]
```diff
@@ -23,78 +23,130 @@ import React from "react";
...
+ * @see Page examples.
+ *
+ * @param {React.ReactNode} [props.children] - a collection of Action components
+ */
+const Actions = ({ children }) => <>{children}</>;
+
+/**
+ * A convenient component representing a Page action
+ *
+ * Built on top of {@link https://www.patternfly.org/components/button PF/Button}
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply web/src/components/core/Page.jsx:47]: Which whitespace? The url is https://www.patternfly.org/components/button. I'm using the `Pf/Button` as label, following the JSDock `@link` tag https://jsdoc.app/tags-inline-link. Maybe is confusing and I should stop using labels

File: web/src/components/storage/ProposalPage.test.jsx [hunk 6]
```diff
@@ -38,6 +36,11 @@ jest.mock("@patternfly/react-core", () => {
 
   };
 });
+// Since Agama sidebar is now rendered by the core/Page component, it's needed
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply web/src/components/storage/ProposalPage.test.jsx:39]: I put them there hoping I can get ride of the mock... but actually at this moment I cannot. Maybe a note in test-utils.js file would be enough. I'll change it. Thanks.

File: web/src/components/core/Page.jsx [hunk 7]
```diff
@@ -135,7 +135,11 @@ const Page = ({ icon, title = "Agama", children }) => {
   // component because the type check.
   //
   // @see https://stackoverflow.com/questions/55729582/check-type-of-react-component
-  const [actions, content] = partition(React.Children.toArray(children), child => child.type === Actions);
+  const [actions, rest] = partition(React.Children.toArray(children), child => child.type === Actions);
+
+  // TODO: move PageOptions to an internal Page component
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply web/src/components/core/Page.jsx]: Done

File: web/src/components/core/Sidebar.test.jsx [hunk 8]
```diff
@@ -21,198 +21,171 @@
...
+  it("renders a link intended for closing it that triggers the onClose callback", async () => {
+    const onClose = jest.fn();
+    const { user } = plainRender(<Sidebar isOpen onClose={onClose} />);
+    const closeLink = screen.getByLabelText(/Hide/i);
+    await user.click(closeLink);
+    expect(onClose).toHaveBeenCalled();
+  });
 });
 
+// NOTE: maybe it's time to kill this feature of keeping the sidebar open
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply web/src/components/core/Sidebar.test.jsx:90]: I think I'll postpone it for a later PR. If so, I'll create an issue right away for not forgetting about it.

File: web/src/components/core/Page.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (imobachgs) [reply web/src/components/core/Page.jsx:57]: No, it is fine, my fault. 🤦🏻

File: web/src/components/core/Page.jsx [hunk 9]
```diff
@@ -120,15 +159,26 @@ const BackAction = () => {
...
+   * To do so, below lines will extract some children based on their type.
+   *
+   * As for actions, the check is straightforward since it is just a convenience
+   * component that consumers will use directly as <Page.Actions>...</Page.Actions>.
+   * However, <Page.Menu> could be wrapped by another component holding all the logic
+   * to build and render an specific menu. Hence, the only option for them at this
+   * moment is to look for children whose type ends in "PageMenu".
+   *
+   * @note: hot reloading could make weird things when working with this
+   * component because the type check.
```
Comment (imobachgs) [inline web/src/components/core/Page.jsx:174]: np: *because of the type check* (?)

**Review by imobachgs — APPROVED** [review]
LGTM. I was thinking about merging *after* releasing version 7 but, to be honest, we can benefit of the extra testing. So, please, feel free to merge.

**Review by imobachgs — APPROVED** [review]

File: web/src/components/core/Page.jsx [hunk 10]
```diff
@@ -23,78 +23,130 @@ import React from "react";
...
+/**
+ * Simple action for navigating back
+ *
+ * @note it will be used by default if a page is mounted without actions
+ *
+ * TODO: Explain below note better
+ * @note that we cannot use navigate("..") because our routes are all nested in
+ * the root.
+ *
+ * @param {React.ReactNode} props.children - content of the action
```
Comment (joseivanlopez) [inline web/src/components/core/Page.jsx:86]: leftover params?

Comment (dgdavid) [reply web/src/components/core/Page.jsx:86]: Yes, I realized a few minutes ago too, while doing a merge into #886 

Will take care of it ASAP.

Comment (dgdavid) [reply web/src/components/core/Page.jsx:86]: Actually something weird happens there (lines switched or so) plus the new props were not documented. As said above, I'll fix it as part of another PR. Or maybe a tiny one for this kind of things, to not forget about them. But tomorrow or next week.

**Comment by joseivanlopez** [comment]
Great work! Almost a canonical PR.

**Pull request merged**